[LLDB] Add unary operators Dereference and AddressOf to DIL

kuilpd opened the pull request.

kuilpd committed:
Add unary operators Dereference and AddressOf

cmtice commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -3,7 +3,9 @@
-expression = primary_expression ;
+expression = unary_expression ;
+
+unary_expression = unary_operator primary_expression ;
```
Need to add just plain primary_expression -- we're not forcing all expressions to have unary ops -- could still be plain variable names.

labath commented:
I see this is just a draft, but in order to save time for everyone, I want to say this as early as possible.

With this patch, the biggest question you will have to answer is "why is the implementation of `&` more complicated than `inner_value->AddressOf()`" (and a similar question for dereferencing). I think I know why you did this (because you want `&variable_of_reference_type` to behave the same way as is C++). I understand that desire, but I don't think this is the way to achieve that. The implementation of `AddressOf` is intentional and it kind of makes sense -- despite of what the C++ standard says, (most) references do have an address, so it makes sense for a debugger to have a way to return that. I think it would also make sense (and I might even prefer that) if `AddressOf` returned the address of the referenced object (the "value" of the reference), but it's not how it works right now -- neither with `AddressOf` nor with the existing implementation of `frame var`. What *doesn't* make sense to me is for `AddressOf` and `frame var` do to something different. If we want to change how this works, the change should probably be inside `AddressOf`, so that it applies the same everywhere. Either way, it will require more discussion. If that's something you want to take on now, then we probably ought to move this to the RFC. If not, then I'd suggest sticking with the status quo (and *maybe* coming back to this later).

I'm also not very convinced by this "FlowAnalysis" thingy. Is it supposed to be a performance optimization (collapsing `*&foo` to `foo`) or does it do something more? I mean, the optimization conceptually makes sense, but I'm somewhat surprised that it makes a difference in practice. Is there a particular pattern you want to optimize with this? I'd be surprised if users actually typed expressions like this.

kuilpd committed:
Fix parsing, remove dereferencing, add unit tests

kuilpd commented:
Trying to add unit tests with this patch. We already have a big test suite that came from `lldb-eval`, and I don't see much reason to rewrite them into Python tests. For now the tests check DIL results against `frame var` (the code to explicitly call them from API is not elegant right now). Later, once we match `frame var` capabilities, we can remove these checks and add checks against LLDB expression evaluator for more complex expressions.

I didn't find an existing way to build the binary for debugging using in-tree compiler and libc++, so had to improvise how to do it in CMake. The compiler options for building I took from Python testing. I think it needs some improvement on how to handle this when the required projects and runtimes are not enabled.
Also, moved the Lexer tests here, but they crash when linked together with other unit tests, not sure why yet.

> With this patch, the biggest question you will have to answer is "why is the implementation of `&` more complicated than `inner_value->AddressOf()`

It's not, the implementation itself is one line, the rest is type and error checking. Same with dereference, only creating an address value first. The node itself is a base for other unary operations in the future.

> What doesn't make sense to me is for AddressOf and frame var to do different things.

I removed the dereferencing so that it matches the current frame var behavior.

labath commented:
> Trying to add unit tests with this patch. We already have a big test suite that came from `lldb-eval`, and I don't see much reason to rewrite them into Python tests. 

> I didn't find an existing way to build the binary for debugging using in-tree compiler and libc++, so had to improvise how to do it in CMake. The compiler options for building I took from Python testing.

This is the reason. Long-term maintainability of the project. We already have (too) many different ways to build test binaries, so you need a very good reason to introduce another method, and frankly, I don't think you have one. If you don't build the test binaries with the same compiler as the rest of the tests, you will expose the test to all of the quirks and bugs of the host compiler, which means that this test will be testing a different thing for every user.

> > With this patch, the biggest question you will have to answer is "why is the implementation of `&` more complicated than `inner_value->AddressOf()`
> 
> It's not, the implementation itself is one line, the rest is type and error checking. Same with dereference, only creating an address value first. The node itself is a base for other unary operations in the future.

The fancy reference handling was my main concern, which you've now addressed. I haven't looked at the rest of the code too closely, so it *may* be fine, but in general, I think you shouldn't be looking at the type of the value in the evaluator: You already have the value in hand, so if the expression says you should dereference it, you should just ask the value to dereference itself. If it (and it's type system) knows how to dereference the value, it will give you a result. If not, you'll get an error.

kuilpd commented:
> This is the reason. Long-term maintainability of the project. We already have (too) many different ways to build test binaries, so you need a very good reason to introduce another method, and frankly, I don't think you have one. 
Is it a general consensus that people shouldn't use unit tests anymore?

> If you don't build the test binaries with the same compiler as the rest of the tests, you will expose the test to all of the quirks and bugs of the host compiler, which means that this test will be testing a different thing for every user.
This is what I tried to do, I looked at how Python builds test binaries and repeated the same commands in CMake.

labath commented:
> Is it a general consensus that people shouldn't use unit tests anymore?

They shouldn't use unit tests for building debug subjects.

kuilpd committed:
Remove unit tests, add Python tests

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -18,6 +18,22 @@
 
 namespace lldb_private::dil {
 
+static lldb::ValueObjectSP
+ArrayToPointerConversion(lldb::ValueObjectSP valobj,
+                         std::shared_ptr<ExecutionContextScope> ctx) {
+  assert(valobj->IsArrayType() &&
+         "an argument to array-to-pointer conversion must be an array");
+
+  uint64_t addr = valobj->GetLoadAddress();
+  llvm::StringRef name = "result";
+  ExecutionContext exe_ctx;
+  ctx->CalculateExecutionContext(exe_ctx);
+  return ValueObject::CreateValueObjectFromAddress(
+      name, addr, exe_ctx,
+      valobj->GetCompilerType().GetArrayElementType(ctx.get()).GetPointerType(),
+      /* do_deref */ false);
+}
+
```
I think this is an interesting question (and one that may deserve a PR of its own). For C, this makes sense, but I'm not sure if we want to do this for all languages. Also, the implementations assumes that an array is just a list of elements (i.e., that the address of the first array element is the same as the address of the array itself). This is true for C, but it may not be true for all languages that may want to support. I'm not sure what's the situation with Swift, but Go (which we don't support, but it makes a good point) has a first class "array slice" type, which is basically equivalent to `llvm::ArrayRef` -- under the hood it's a pointer+length pair. I think it'd be reasonable for Go's type system to claim that this is an array type, but you definitely couldn't get to the first element of the slice by `reinterpret_cast`ing the address of the slice.

Another problem here is that this ignores any data formatters the type may have. A user may have defined a synthetic child provider for the array type, and that provider may define (by providing a special `$$dereference$$` child) what it wants to be the result "dereferencing" that object. By special-casing array types, you're subverting that behavior.

For these reasons, I think this logic should go into `ValueObject::Dereference`. There, we could either:
- implement this logic directly (but without reinterpret_casting -- just do something roughly equivalent to `GetChildAtIndex(0)`)
- or delegate the operation to the type system in some way, so that it can define what exactly (if anything) does "dereferencing an array" mean for the language in question

@jimingham, @adrian-prantl, do you concur, or have anything to add here?

jimingham replied:
Yes, I concur.  As a general rule, we want the evaluator to know as little as possible about how any operation is performed and delegate as much as possible to ValueObject methods.  

The evaluator can't know anything about the language of the values it is operating on or it will either only work for C-backed ValueObjects or end up cluttered with "if C/C++/swift/go/future_language" code, which we really want to avoid.  However, ValueObjects always know what type system their type comes from, and so they can naturally do the right thing for their language.  So we should defer to them whenever we're doing any non-trivial operations.

labath replied:
Well.. that's why I said it's an interesting question :)

I don't think it should be a language check. It should be plumbed into the type system code.. somehow. Currently `ValueObject::Dereference` calls `GetCompilerType().GetChildCompilerTypeAtIndex` (which eventually calls into the type system). That's kind of a problem because the type system doesn't actually know that it's being asked to dereference. Nonetheless, after it returns, it does tell you whether the returned child is a dereference of the parent (via the `child_is_deref_of_parent` by-ref result).

So, one option that *might* be acceptable (to me, others may disagree) would be to call this function on array types as well and then to check whether the result has the `child_is_deref_of_parent` flag set (we'd also need to change the TypeSystemClang implementation so that it sets the flag for the first element of the array).

The alternative would be to create a new function that explicitly tells the type system that you're trying to dereference the value, and then it can do pretty much anything it wants.

jimingham replied:
Is there any downside to the second option?  It seems to me the general philosophy here is to have DIL do as little as possible and the type system as much as possible, since then we're asking the entity that actually knows what it is doing...

labath replied:
Yes, that's definitely a job for a separate patch.

To answer Jim's question, I don't see a specific downside to the second option. Just some open questions. I don't exactly know what the interface of the new methods should be for instance (maybe it should not return a ValueObject since GetChildCompilerTypeAtIndex does not either). Ideally, I also wouldn't want too much code duplication between this and GetChildCompilerTypeAtIndex. But generally, yes, I think this would be better.

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -232,4 +263,105 @@ Interpreter::Visit(const IdentifierNode *node) {
+    return rhs;
+  }
+  case UnaryOpKind::AddrOf: {
+    if (node->rhs()->is_rvalue()) {
```
While I don't feel as strongly about this as about the array dereference thing, I do think that (and I hope you start to see a theme here) we shouldn't need to track the rvalue-ness of an expression/value separately from the ValueObject itself. I think that a ValueObject should know whether it's able to take the address of itself, because people may invoke `AddressOf` directly. The current behavior of AddressOf doesn't make much sense to me -- it lets me chain AddressOf twice (but not three times!) -- but I think that's a bug that should be fixed inside that function, not worked around in the DIL (I'm not saying you should fix that here -- you can leave that to a separate patch, or even a separate person):
```
(lldb) script lldb.frame.FindVariable("y")
(int) y = 47
(lldb) script lldb.frame.FindVariable("y").AddressOf()
(int *) &y = 0x00007fffffffd7d0
(lldb) script lldb.frame.FindVariable("y").AddressOf().AddressOf()
(int **) &&y = 0x00007fffffffd7d0 # ???
(lldb) script lldb.frame.FindVariable("y").AddressOf().AddressOf().Dereference()
(int *) *&&y = 0x000000000000002f # :facepalm: 
(lldb) script lldb.frame.FindVariable("y").AddressOf().AddressOf().AddressOf()
No value
(lldb) script lldb.frame.FindVariable("y").AddressOf().AddressOf().AddressOf().GetError()
error: error: invalid value object
```

kuilpd replied:
From what I could gather, the first one is `ValueObject::AddressOf` that creates a `ValueObjectConstResult` and fills in its address. The second one calls `ValueObjectConstResultImpl::AddressOf` that also creates `ValueObjectConstResult` but with no address this time (which makes sense for a const value), so the third one fails.
I'm guessing that the first `ValueObjectConstResult` is pretty much an implementation of a reference, so it has to have an address behind it so it can be dereferenced normally.

kuilpd replied:
I created #137688 to address this issue.

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -232,4 +263,105 @@ Interpreter::Visit(const IdentifierNode *node) {
+      return llvm::make_error<DILDiagnosticError>(m_expr, errMsg,
+                                                  node->GetLocation());
+    }
+    if (rhs->IsBitfield()) {
+      return llvm::make_error<DILDiagnosticError>(
+          m_expr, "address of bit-field requested", node->GetLocation());
+    }
```
What happens if you call `AddressOf` on a bitfield?

kuilpd replied:
`ValueObject::GetAddressOf` returns `LLDB_INVALID_ADDRESS` for a bitfield without an error, so this check is to provide a meaningful error message.

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -232,4 +263,105 @@ Interpreter::Visit(const IdentifierNode *node) {
+  assert(rhs->GetCompilerType().IsPointerType() &&
+         "invalid ast: must be a pointer type");
+
+  if (rhs->GetDerefValobj())
```
Why doesn't this call `Dereference()`?

`GetDerefValobj` is currently unused. It was added in #87197, but without any context. Such a function *might* be useful in some situations, but I don't think this is one of them. You really should let the ValueObject tell you what is the result of dereferencing it.

kuilpd replied:
This checks if the value has already been dereferenced. The same check is also done in the beginning of `Dereference()`, but if we want to skip `CreateValueObjectFromAddress` as well, this is done separately beforehand.

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -232,4 +263,105 @@ Interpreter::Visit(const IdentifierNode *node) {
+  ExecutionContext exe_ctx(m_target.get(), false);
+  lldb::ValueObjectSP value = ValueObject::CreateValueObjectFromAddress(
+      name, base_addr, exe_ctx, pointer_type,
+      /* do_deref */ false);
```
This is the comment style used in llvm

```suggestion
      /*do_deref=*/false);
```

labath commented on lldb/source/ValueObject/DILEval.cpp:
```diff
@@ -232,4 +263,105 @@ Interpreter::Visit(const IdentifierNode *node) {
+  return value->Dereference(error);
+}
+
+} // namespace lldb_private::dil
```
missing newline

labath commented on lldb/include/lldb/ValueObject/DILAST.h:
```diff
@@ -81,6 +90,27 @@ class IdentifierNode : public ASTNode {
+
+private:
+  UnaryOpKind m_kind;
+  ASTNodeUP m_rhs;
```
It's kinda weird to have the *right* side of a unary expression (and it would get particularly weird if we end up having a postfix `++` operator). Clang calls this `Val` or SubExpr`, so maybe call this m_(sub_)expr?

labath reviewed (commented):
> The fancy reference handling was my main concern, which you've now addressed. I haven't looked at the rest of the code too closely, so it may be fine, but in general, I think you shouldn't be looking at the type of the value in the evaluator: You already have the value in hand, so if the expression says you should dereference it, you should just ask the value to dereference itself. If it (and it's type system) knows how to dereference the value, it will give you a result. If not, you'll get an error.

I took a closer look at the implementation now, and I do think you're looking at the type of the value too much. More on that in the inline comments.

I'm also still waiting on the rationale behind the FlowAnalysis subsystem.

kuilpd commented on lldb/docs/dil-expr-lang.ebnf:
```diff
@@ -3,7 +3,12 @@
-expression = primary_expression ;
+expression = unary_expression ;
+
+unary_expression = unary_operator expression
```
We have a final grammar ready, but it will have to be adjusted every patch anyway. For now, this allows to write multiple unary operators in a row, like `&*foo`.

kuilpd commented:
> I'm also not very convinced by this "FlowAnalysis" thingy. Is it supposed to be a performance optimization (collapsing *&foo to foo) or does it do something more?

FlowAnalysis is also used in subscript operator, to short-circuit getting an address of an element: `&arr[1]`, which could be a much more common use case.

labath commented:
> > I'm also not very convinced by this "FlowAnalysis" thingy. Is it supposed to be a performance optimization (collapsing *&foo to foo) or does it do something more?
> 
> FlowAnalysis is also used in subscript operator, to short-circuit getting an address of an element: `&arr[1]`, which could be a much more common use case.

I see. So, I agree `&a[i]` is likelier than `&*a`. However, I find it hard to imagine how to do that correctly without making too many C-like assumptions about what is an "array". To do this correctly, I think the logic would have to be moved inside the type system, which will be a lot of work for possibly very little gain -- I'm also rather sceptical that something like this will be appreciably faster than just performing the operations in the order prescribed by the expression.

I would really want to suggest that you take out this feature and implement dereference/addressof in the simplest possible terms. Once we have all of the pieces in place (*), we can take a look the performance of the code and see if anything needs optimizing. It may turn out that that we can optimize something else so that `[i]` is fast and doesn't need to be optimized by addressof collapsing. Or it may not be a problem at all. I'm sure lldb-eval had its reasons to implement this, but its use case was very different from this, and it wasn't in a position to change lldb to make things faster. We are.

(*) My goal is to replace/delete the old "frame var" implementation as soon as possible. Until we do that, every line here is just adding technical debt. Fancy addressof handling is not required to do that. What we need is (simple) implementations of operations that are supported currently. Besides `&` and `*`, that's `.`, `->` and `[]`.

kuilpd committed:
Remove array conversion, cut down evaluation of & and *.

werat commented:
> I'm sure lldb-eval had its reasons to implement this, but its use case was very different from this, and it wasn't in a position to change lldb to make things faster. We are.

As far as I remember flow analysis was implemented because according to C rules `&*p` is equivalent to `p`, which is important if `p` is NULL or "invalid". Same goes for array dereferencing. 

I could see how DIL may not want to care about this and just do dereference and return error if the pointer is invalid. I had to implement it in lldb-eval because we did have some user expressions that relied on this fact :)

kuilpd committed:
Rename rhs to operand, remove extra check and return `ValueObject::Dereference` error, cleanup code.

kuilpd commented:
@labath 
Should I move the check for `void *` to `TypeSystemClang::GetDereferencedType` in the other PR?

labath commented:
Yes, I think that would be best, although if you structure the error message like the current command, it may not be even necessary to have a dedicated error message for that. This is what I get with the current implementation, and I think it's quite okay:
```
(lldb) v *x.x->x->x->x->v
error: dereference failed: (void *) x.x->x->x->x->v
```

kuilpd committed:
Remove a pointer to void check
